yamllint/rules/quoted_strings.py

```diff
@@ -259,9 +269,16 @@ def check(conf, token, prev, next, nextnext, context):
 
     elif conf['required'] == 'only-when-needed':
 
+        # Quotes needed on strings containing flow tokens
+        if context['flow_cnt'] > 0 and \
+           {',', '[', ']', '{', '}'} & set(token.value):
+            has_flow_token = True
+        else:
+            has_flow_token = False
+
         # Quotes are not strictly needed here
         if (token.style and tag == DEFAULT_SCALAR_TAG and token.value and
-                not _quotes_are_needed(token.value)):
+                not has_flow_token and not _quotes_are_needed(token.value)):
```

Comment:
It would make more sense / be more readable to keep only `_quotes_are_needed()` here.

I propose to change it a bit:
```suggestion
                not _quotes_are_needed(token.value, context['flow_cnt'] > 0)):
```
and
```python
def _quotes_are_needed(string, is_inside_a_flow):
    # check for ',', '[', ']', '{', '}' here if is_inside_a_flow
```